src/main/java/com/google/sps/data/LearningPathService.java

```diff
@@ -136,16 +152,73 @@ public LearningItem loadItem(long itemId) throws EntityNotFoundException {
 		return mapEntityToLearningItem(itemId, item);
 	}
 
-	private LearningItem mapEntityToLearningItem(long sectionId, Entity e) {
+    public LearningItem loadFeedbackItems(long itemId) throws EntityNotFoundException {
+		Entity item = datastore.get(KeyFactory.createKey(LEARNING_ITEM, itemId));
+		return mapEntityToLearningItem(itemId, item);
+	}
+    private LearningItem mapEntityToLearningItem(long sectionId, Entity e) {
 		return new LearningItem(e);
 	}
 
+    public List<ItemFeedback> loadItemFeedbacks(long learningPathId) throws EntityNotFoundException {
+		Query query = new Query(ITEM_FEEDBACK).addSort("id")
+				.setFilter(new Query.FilterPredicate("learningPath", Query.FilterOperator.EQUAL, learningPathId));
+
+		List<Entity> items = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+
+		return items.stream().map(e -> mapEntityToItemFeedback(learningPathId, e)).collect(Collectors.toList());
+	}
+
+    public List<ItemFeedback> loadItemFeedbacksUser(long userID) throws EntityNotFoundException {
+		Query query = new Query(ITEM_FEEDBACK).addSort("id")
+			.setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userID));
+
+		List<Entity> items = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+
+		return items.stream().map(e -> mapEntityToItemFeedback(userID, e)).collect(Collectors.toList());
+	}
+
+    private ItemFeedback mapEntityToItemFeedback(long sectionId, Entity e) {
+		return new ItemFeedback(e);
+	}
+
+
+    public long getSectionCompletion(long userId, long sectionId ,long learningPath)throws EntityNotFoundException { 
+       // filter 
+        List<ItemFeedback> feedbacks = loadItemFeedbacksUser(userId).stream()
+                                    .filter( f -> f.isCompleted()).collect(Collectors.toList());
+        List<LearningItem> itemsInSection = loadItems(sectionId);
+
+        List<Long> feedbackId = feedbacks.stream().map(p -> p.getLearningItem())
+                                .collect(Collectors.toList()); 
+
+        List<Long> sectionIds = itemsInSection.stream().map(p -> p.getId())
+                                .collect(Collectors.toList()); 
+       
+        feedbackId.retainAll(sectionIds);
+
+        long proportion = feedbackId.size()/itemsInSection.size();
+       
+        return proportion;
+    }
+
+    public long getLearningPathCompletion(long userId, long learningPathId)throws EntityNotFoundException {
+        LearningPath path = load(learningPathId);
```

Comment:
Note that when we load the learning path, we recursively fetch all the sections and items.

We could pass the `LearningSection` into the `getSectionCompletion` method and avoid having to load the `itemsInSection` a second time